File: doheeahn/3주차/PGS_ 피로도.java
class Solution {
    static int result =0;
    public int solution(int k, int[][] dungeons) {
        int answer = -1;
        boolean[] visited = new boolean[dungeons.length];
        perm(k,dungeons,visited,0,0);
        
        answer = result;
        return answer;
    }
    
    public void perm(int k, int[][] dungeons, boolean[] visited,int cnt,int max){
        if(cnt == dungeons.length){
            if(result < max){
                result = max;
            }
            return;
        }
        
        for(int i=0;i<dungeons.length;i++){
            if(visited[i]){
                continue;
            }
            visited[i] = true;
            if(dungeons[i][0]>k){
                perm(k,dungeons,visited,cnt+1,max);
            }
            else{
                perm(k-dungeons[i][1],dungeons,visited,cnt+1,max+1);
            }
            visited[i] = false;
        }
    }
}

File: daehwa_park/2주차/PGS_피로도.py
def solution(k, dungeons):
    answer = -1

    visited = [False for _ in range(len(dungeons))]

    answer = perm(0, len(dungeons), k, visited, dungeons)

    return answer


def perm(cnt, dst, pi, visited, dungeons):  # cnt : 선택한 개수, dst : 최대 가능 개수, pi : 현재 피로도
    if cnt == dst:
        return cnt

    max_ans = cnt
    for i in range(dst):
        if not visited[i] and pi >= dungeons[i][0]:
            visited[i] = True
            max_ans = max(max_ans, perm(cnt + 1, dst, pi - dungeons[i][1], visited, dungeons))
            visited[i] = False
    return max_ans
